_posts/2022-04-22-switch-to-java-11.adoc

```diff
@@ -0,0 +1,138 @@
+So it's time to start thinking about using a more recent JDK, and the new major Debezium release is a natural milestone when to do the switch.
+
++++<!-- more -->+++
+
+Since early releases of Debezium 2.0, hopefully from the very first one `Alpha1`, Debezium bits will be compiled to Java 11 byte code.
+Therefore, Java 11 will be required to run Debezium in the next major update.
+Also, if you use any of the Debezium bits in your project, you will have to switch to Java 11.
+
+But wait, what does Java 11/17 in the title mean?
+Is it there just to scare you or we are going to actually switch to Java 17?
+
+<dramatic pause here>
+
+No, we don't want to scare you.
+We are actually planning to switch to Java 17, but only for the test suite.
+This will allow us to use more recent Java features in the tests, while users still can stay on Java 11.
+
+== Implementation
+
+Setting different byte code level for code and tests is pretty easy with Maven, 
+you just need to set following system properties:
+
+[source, xml]
+----
+<maven.compiler.release>11</maven.compiler.release>
+<maven.compiler.testRelease>17</maven.compiler.testRelease>
+----
+
+After switching to Java 11, checkstyle plugin and impsort plugin (plugin which takes care about proper order of imports) started to fail.
+However, bumping their versions to the latest releases has solved all the issues.
+
+This was the easy part.
+The most difficult part was the Cassandra connector.
+
+=== Cassandra connector
+
+Cassandra connector now provides support for Cassandra 3 as well as for Cassandra 4.
+Cassandra 4 https://cassandra.apache.org/doc/4.0/cassandra/new/java11.html[works with Java 11], 
+but running Cassandra 3 with Java 11 is not possible (or at least requires some hacking).
+Current test implementation doesn't run Cassandra in a container as we do in tests for other DB connectors, but runs Cassandra in embedded mode.
+Therefore if you want to run the tests with Java 11 (or 17), tests for Cassandra 3 module would fail.
+Obvious solution is to run Cassandra in a container with Java 8.
+This sounds good, but this approach has one pitfall.
+The Cassandra connector needs access to Cassandra log files as it reads the dedicated log file to obtain CDC events, so the tests need to access Cassandra files in the container.
+This can be solved quite easily using a temporary directory, .one in e.g `target`, giving access to all users and moutable within the container.
+Cassandra running in the container can later on use this mounted volume for storing its data.
+The real issue starts when you try to do the cleanup after the tests.
+As Cassandra runs under dedicated user `cassandra` in the container, which is very likely not present on the test machine (or with different UID/GID), 
+cleanup fails when it tries to delete the temporary directory with Cassandra files.
+These files were created by a different user with insufficient rights for other users to delete them.
```

Comment:
I don't quite get the problem yet. The log files are created by the `cassandra` user within the container, right? So why is it that user cannot delete the files then?

Reply:
Files were created by `cassandra` user withing the containers. But the directory where they were create is bind mount from temporary dir created by test in `target` directory, so these files are not in Docker FS overlay, but directly in project `target` directory (as tests need the access to the cassadra log files). And user running the tests typically don't have sufficient right to delete files created by different user. Is it more clear now?

I tried to explain it more also in the text.